pkg/reconciler/kubernetes/adapter.go

```diff
@@ -91,93 +125,352 @@ func (g *kubeClientAdapter) Kubeconfig() string {
+	restConfig.ContentConfig = resource.UnstructuredPlusDefaultContentConfig()
+	restConfig.GroupVersion = &gv
+
+	if len(gv.Group) == 0 {
+		restConfig.APIPath = "/api"
+	} else {
+		restConfig.APIPath = "/apis"
+	}
+
+	return rest.RESTClientFor(&restConfig)
+}
+
+func (g *kubeClientAdapter) deleteResource(infoTarget *resource.Info) (*Resource, error) {
+	result, errs := g.helmClient.Delete(kube.ResourceList{infoTarget})
+	if errs != nil || result == nil || result.Deleted == nil {
+		g.logger.Warnf("kubeClient failed to delete %s '%s' (namespace: %s): %v",
+			infoTarget.Object.GetObjectKind().GroupVersionKind().Kind, infoTarget.Name, infoTarget.Namespace, errs)
+		return nil, errs[0]
+	}
+	infoDeleted := result.Deleted.Get(infoTarget)
+	if infoDeleted == nil {
+		g.logger.Errorf("Deleted Resource mismatch: Target: %s '%s' (namespace: %s) - Actual:  %s '%s' (namespace: %s)",
+			infoTarget.Object.GetObjectKind().GroupVersionKind().Kind, infoTarget.Name, infoTarget.Namespace,
+			infoDeleted.Object.GetObjectKind().GroupVersionKind().Kind, infoDeleted.Name, infoDeleted.Namespace)
+		return nil, fmt.Errorf("Deleted Resource mismatch")
+	}
+	g.logger.Debugf("kubeClient delete %s '%s' (namespace: %s) successfully.",
+		infoDeleted.Object.GetObjectKind().GroupVersionKind().Kind, infoDeleted.Name, infoDeleted.Namespace)
+	return &Resource{
+		Kind:      infoDeleted.Object.GetObjectKind().GroupVersionKind().Kind,
+		Name:      infoDeleted.Name,
+		Namespace: infoDeleted.Namespace,
+	}, nil
+}
+
+func (g *kubeClientAdapter) deployResource(infoOriginal, infoTarget *resource.Info, namespaceOverride string) (*Resource, error) {
+
+	helper := resource.NewHelper(infoTarget.Client, infoTarget.Mapping)
+	setNamespaceIfScoped(namespaceOverride, infoTarget, helper)
+
+	strategy, err := newDefaultUpdateStrategyResolver(helper).Resolve(infoTarget)
+	if err != nil {
+		return nil, err
+	}
+
+	if strategy == SkipUpdateStrategy {
+		return nil, nil
+	}
+
+	//retry the reconciliation in case of an error
+	err = retry.Do(g.deployResourceFunc(infoOriginal, infoTarget, strategy),
+		retry.Attempts(uint(g.config.MaxRetries)),
+		retry.Delay(g.config.RetryDelay),
+		retry.LastErrorOnly(false),
+		retry.Context(context.Background()))
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "kubeClient failed to update %s '%s' (namespace: %s)",
+			infoTarget.Object.GetObjectKind().GroupVersionKind().Kind, infoTarget.Name, infoTarget.Namespace)
+	}
+
+	resource := &Resource{
+		Kind:      infoTarget.Object.GetObjectKind().GroupVersionKind().Kind,
+		Name:      infoTarget.Name,
+		Namespace: infoTarget.Namespace,
+	}
+	return resource, nil
+}
+
+func (g *kubeClientAdapter) deployResourceFunc(infoOriginal, infoTarget *resource.Info, strategy UpdateStrategy) func() error {
+	return func() error {
+		replaceResource := strategy == ReplaceUpdateStrategy
+		_, err := g.helmClient.Update(kube.ResourceList{infoOriginal}, kube.ResourceList{infoTarget}, replaceResource)
```

Comment:
If you read the implementation of this helmClient.Update, at the last step, it will delete all resources which are in original resources but not in target resources. That is the reason why here I just use a single resource to do update, to prevent it start that delete resources process.